public/app/features/alerting/unified/utils/notification-policies.ts

```diff
@@ -11,111 +11,84 @@ import { Labels } from 'app/types/unified-alerting-dto';
+// So we cannot use null as an indicator of no match
 interface LabelMatchResult {
   match: boolean;
-  matchers: ObjectMatcher[];
+  matcher: ObjectMatcher | null;
 }
 
+type LabelsMatch = Map<Label, LabelMatchResult>;
+
 interface MatchingResult {
   matches: boolean;
-  details: Map<ObjectMatcher, Label[]>;
-  labelsMatch: Map<Label, LabelMatchResult>;
+  labelsMatch: LabelsMatch;
 }
 
-// check if every matcher returns "true" for the set of labels
-function matchLabels(matchers: ObjectMatcher[], labels: Label[]): MatchingResult {
-  const details = new Map<ObjectMatcher, Label[]>();
-
-  // If a policy has no matchers it still can be a match, hence matchers can be empty and match can be true
-  // So we cannot use empty array of matchers as an indicator of no match
-  const labelsMatch = new Map<Label, { match: boolean; matchers: ObjectMatcher[] }>(
-    labels.map((label) => [label, { match: false, matchers: [] }])
-  );
+// returns a match results for given set of matchers (from a policy for instance) and a set of labels
+export function matchLabels(matchers: ObjectMatcher[], labels: Label[]): MatchingResult {
+  const matches = matchLabelsSet(matchers, labels);
 
-  const matches = matchers.every((matcher) => {
-    const matchingLabels = labels.filter((label) => isLabelMatch(matcher, label));
+  // create initial map of label => match result
+  const labelsMatch: LabelsMatch = new Map(labels.map((label) => [label, { match: false, matcher: null }]));
 
-    matchingLabels.forEach((label) => {
-      const labelMatch = labelsMatch.get(label);
-      // The condition is just to satisfy TS. The map should have all the labels due to the previous map initialization
-      if (labelMatch) {
-        labelMatch.match = true;
-        labelMatch.matchers.push(matcher);
-      }
-    });
+  // for each matcher, check which label it matched for
+  matchers.forEach((matcher) => {
+    const matchingLabel = labels.find((label) => isLabelMatch(matcher, label));
 
-    if (matchingLabels.length === 0) {
-      return false;
+    // record that matcher for the label
+    if (matchingLabel) {
+      labelsMatch.set(matchingLabel, {
+        match: true,
+        matcher,
+      });
     }
-
-    details.set(matcher, matchingLabels);
-    return matchingLabels.length > 0;
   });
 
-  return { matches, details, labelsMatch };
+  return { matches, labelsMatch };
+}
+
+// Compare set of matchers to set of label
+export function matchLabelsSet(matchers: ObjectMatcher[], labels: Label[]): boolean {
+  for (const matcher of matchers) {
+    if (!isLabelMatchInSet(matcher, labels)) {
+      return false;
+    }
+  }
+  return true;
 }
 
 export interface AlertInstanceMatch {
   instance: Labels;
-  matchDetails: Map<ObjectMatcher, Labels>;
-  labelsMatch: Map<Label, LabelMatchResult>;
+  labelsMatch: LabelsMatch;
 }
 
 export interface RouteMatchResult<T extends Route> {
   route: T;
-  details: Map<ObjectMatcher, Label[]>;
```

Comment:
We've decided to remove the `details` here since it's not being used and adds a bit of complexity we can avoid